hedera-node/hedera-app/src/main/java/com/hedera/node/app/roster/schemas/V0540RosterSchema.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.roster.schemas;
+
+import com.hedera.hapi.node.base.SemanticVersion;
+import com.hedera.hapi.node.state.primitives.ProtoBytes;
+import com.hedera.hapi.node.state.roster.Roster;
+import com.hedera.hapi.node.state.roster.RosterState;
+import com.swirlds.state.spi.MigrationContext;
+import com.swirlds.state.spi.Schema;
+import com.swirlds.state.spi.StateDefinition;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Set;
+
+/**
+ * Roster Schema
+ */
+public class V0540RosterSchema extends Schema {
+    public static final String ROSTER_KEY = "ROSTER";
+    public static final String ROSTER_STATES_KEY = "ROSTER_STATE";
+    /** this can't be increased later so we pick some number large enough, 2^16. */
+    private static final long MAX_ROSTERS = 65536L;
```

Comment:
I'm a bit concerned about this number. If we consider the DAB is implemented and rosters indeed change once a day, this upper limit allows the map to store ~180 years worth of rosters, which is almost two centuries. I don't believe it's reasonable to store that many entries in this map for all practical purposes. In two centuries, I believe this map implementation will support dynamic resizing (provided it still exists by that time.)

On the flip side, this maximum limit may define sizes of some resources (RAM, disk, and whatnot) being allocated by the implementation today. And this extra space would simply be wasted because there's no way for us to populate that many rosters in any reasonable time.

I suggest to drastically reduce this size. I think that 256, while still pretty large in my opinion, would still be a lot more reasonable than the 64K value you're using here.

Reply:
See https://github.com/hashgraph/hedera-services/pull/14861#discussion_r1720307951

Reply:
Yes, I saw that. But my point still stands: I'm not convinced that 64K is a good number per my arguments above.

Reply:
I'm not convinced any number > 0 is a good or bad number until i fully understand how this is used, which i can't seem to find (hence why I blindly accepted @rbair23's suggestion). @artemananiev any pointers to how this number is used?

Reply:
This `maxNumberOfKeys` hint is used to calculate the size of "keys to paths" bucket index in MerkleDb. If the hint is 64K, index size will be 2K, which is 16K bytes (every index entry is a long)

Reply:
Note that with Virtual Mega Map effort, all these hints will become obsolete anyway

Reply:
Thanks for the details. While 16KB doesn't sounds like much, I still feel that having an index of 2K entries for this particular collection doesn't seem reasonable based on the arguments that I've provided above.

I'm assuming that we divide the hint value by 32 to determine the size of the index.

With this assumption in mind, I suggest to set the hint to 8192. I think this number is large enough to support any reasonable use-cases for this map. And at the same time it's small enough to result in an index of 256 entries/2KB in size which is more than enough for any immediate and mid-term use-cases.

Since the hints will be deprecated, the value we choose today wouldn't really matter in the long run.

Reply:
we agreed to leave this as-is.